### PR TITLE
feat(trailhead): add support for multiple newletters

### PR DESCRIPTION
**Pull request by vbudhram**
Opening this PR up as a draft for https://github.com/mozilla/fxa/issues/1102.

This has added initial support for subscribing to multiple email newsletters. When a user verifies an account via `/verify_code` endpoint, they can pass an additional array `newsletters` that contains the newsletter slug names defined in https://docs.google.com/spreadsheets/d/1CdsuwXBhXhJ9LBMthRMDx9ATEGWDiKXlZoPFWLjjKzI/edit#gid=0. 

- [x] Update tests
- [x] Upate JS client
- [x] Verfiy correct metrics emitted

### CODE REVIEW

File: packages/fxa-auth-server/test/local/metrics/amplitude.js [hunk 1]
```diff
@@ -400,6 +400,67 @@ describe('metrics/amplitude', () => {
       });
     });
 
+    describe('account.verified, newsletters is undefined', () => {
```
Comment (vbudhram) [inline packages/fxa-auth-server/test/local/metrics/amplitude.js]: These tests will fail since auth-server needs a new taged fxa-shared version.

**Comment by pmac** [comment]
Basket will simply filter out any newsletter ID with a `false` value, and will not cause an unsubscribe event to happen if the user is already subscribed to that newsletter. All basket needs is a list of the IDs, so you could just send that if you'd rather. Otherwise I can just ignore the ones that are `false`.

**Comment by pmac** [comment]
Also, I'm told that basket should expect a `country` field as well. Is that happening in this PR or another?

**Comment by vbudhram** [comment]
> and will not cause an unsubscribe event to happen if the user is already subscribed to that newsletter.

Oh interesting, I was under the impression that it would get unsubscribed. If that is the case I'll update this to send a list of IDs.

> Also, I'm told that basket should expect a country field as well. Is that happening in this PR or another?

This was added in https://github.com/mozilla/fxa/pull/1051, however it looks like I missed updating the the `notifyAttachedServices` in the `/verify_email` endpoint. I'll prep a new PR for that.

**Comment by pmac** [comment]
> Oh interesting, I was under the impression that it would get unsubscribed. If that is the case I'll update this to send a list of IDs.

I can make that happen if required, but normally these opt-in things are just "select the ones you want" and not so much "manage your preferences". Depends on the presentation I guess. If we do need to update newsletter preferences instead of just do some opt-in please let me know.

**Comment by pmac** [comment]
> This was added in #1051, however it looks like I missed updating the the `notifyAttachedServices` in the `/verify_email` endpoint. I'll prep a new PR for that.

If I'm reading that PR correctly it looks like it's adding the country name to the data, whereas we need the 2 character country code. I don't think it would be reliable for me to change the name back to the code on the basket side.

**Comment by pmac** [comment]
I've talked with the email folks and they confirmed that we should consider people who've checked the boxes for any of those 3 newsletters to be "opt-in"s but if they don't check any that they're not unsubscribing. Chances are very slim that they'd have ever had an opportunity to subscribe to any of those anywhere else regardless. But the place where they manage their subscriptions to opt-out of any of them is the "Email Preferences" page on www.m.o which they can access via their FxA preferences.

File: packages/fxa-auth-server/lib/routes/emails.js [hunk 2]
```diff
@@ -279,14 +279,21 @@ module.exports = (log, db, mailer, config, customs, push, verificationReminders)
             reminder: isA.string().regex(REMINDER_PATTERN).optional(),
             type: isA.string().max(32).alphanum().optional(),
             style: isA.string().allow(['trailhead']).optional(),
-            marketingOptIn: isA.boolean()
+            marketingOptIn: isA.boolean(),
+            newsletters: isA.array().items(isA.string().valid(
```
Comment (shane-tomlinson) [inline packages/fxa-auth-server/lib/routes/emails.js]: :+1: to this change.

File: packages/fxa-auth-server/lib/routes/emails.js [hunk 3]
```diff
@@ -279,14 +279,21 @@ module.exports = (log, db, mailer, config, customs, push, verificationReminders)
...
+              'knowledge-is-power',
+              'test-pilot',
+              'take-action-for-the-internet',
+              'firefox-accounts-journey'
```
Comment (shane-tomlinson) [inline packages/fxa-auth-server/lib/routes/emails.js]: nit: Can you alphabetize the list?

File: packages/fxa-auth-server/lib/routes/emails.js [hunk 4]
```diff
@@ -279,14 +279,21 @@ module.exports = (log, db, mailer, config, customs, push, verificationReminders)
...
+              )).optional()
           }
         }
       },
       handler: async function (request) {
         log.begin('Account.RecoveryEmailVerify', request);
 
-        const { code, marketingOptIn, reminder, service, type, uid, style } = request.payload;
+        const { code, marketingOptIn, newsletters, reminder, service, style, type, uid } = request.payload;
```
Comment (shane-tomlinson) [inline packages/fxa-auth-server/lib/routes/emails.js]: Is the default value of `newsletters` if none are passed in an empty array?

File: packages/fxa-auth-server/test/local/metrics/amplitude.js [hunk 5]
```diff
@@ -400,6 +400,44 @@ describe('metrics/amplitude', () => {
...
+
+      it('did not call log.error', () => {
+        assert.equal(log.error.callCount, 0);
+      });
+
+      it('called log.amplitudeEvent correctly', () => {
+        assert.equal(log.amplitudeEvent.callCount, 1);
+        const args = log.amplitudeEvent.args[0];
+        assert.equal(args[0].event_type, 'fxa_reg - email_confirmed');
+        assert.equal(args[0].user_properties.newsletters, undefined);
```
Comment (shane-tomlinson) [inline packages/fxa-auth-server/test/local/metrics/amplitude.js]: assert.equal against undefined could match any falsey value, can't it? Should you use assert.isUndefined here? Or, should we pass an empty array?

Comment (philbooth) [reply packages/fxa-auth-server/test/local/metrics/amplitude.js]: Fwiw I think we need to pass an empty array for this case, because of the way `user_properties` are carried forward over time. Otherwise the absence of any value will just revert to whatever the previous value was.

Comment (philbooth) [reply packages/fxa-auth-server/test/local/metrics/amplitude.js]: (sorry @vbudhram, I realise now I should have mentioned that in the `fxa-shared` PR)

Comment (philbooth) [reply packages/fxa-auth-server/test/local/metrics/amplitude.js]: Actually, looking at the current code in `fxa-shared`, it will be fine as long as it gets passed in as an empty array in the fully-unsubscribed case. Worth having an extra test here that passes it in as an empty array?

Comment (philbooth) [reply packages/fxa-auth-server/test/local/metrics/amplitude.js]: (and none of this is relevant to Shane's original comment, sorry; I agree that `isUndefined` is a better assertion to use fwiw)

Comment (shane-tomlinson) [reply packages/fxa-auth-server/test/local/metrics/amplitude.js]: > (and none of this is relevant to Shane's original comment, sorry; I agree that isUndefined is a better assertion to use fwiw)

It's all relevant! Passing an empty array makes sense to me too.

Comment (vbudhram) [reply packages/fxa-auth-server/test/local/metrics/amplitude.js]: Sold! Updated this to pass an empty array.

File: packages/fxa-auth-server/lib/routes/emails.js [hunk 6]
```diff
@@ -279,14 +279,21 @@ module.exports = (log, db, mailer, config, customs, push, verificationReminders)
             reminder: isA.string().regex(REMINDER_PATTERN).optional(),
             type: isA.string().max(32).alphanum().optional(),
             style: isA.string().allow(['trailhead']).optional(),
-            marketingOptIn: isA.boolean()
+            marketingOptIn: isA.boolean(),
+            newsletters: isA.array().items(isA.string().valid(
+              'firefox-accounts-journey',
+              'knowledge-is-power',
+              'test-pilot',
+              'take-action-for-the-internet',
```
Comment (shane-tomlinson) [inline packages/fxa-auth-server/lib/routes/emails.js]: nit: take-action and test-pilot are out of order
